### PR TITLE
libct: fix a race with systemd removal

### PR DESCRIPTION
**Pull request by kolyshkin**
For the previous attempt to fix that (and added test cases), see commit 9087f2e827d971 (PR #2338).

Alas, it's not always working because of cgroup directory TOCTOU.

To solve this and avoid the race, add an error _after_ the operation. Implement it as a method that ignores the error that should be ignored. Instead of currentStatus(), use faster runType(), since we are not interested in Paused status here.

For Processes(), remove the pre-op check, and only use it after getting an error, making the non-error path more straightforward.

For Signal(), add a second check after getting an error. The first check is left as is because signalAllProcesses might print a warning if the cgroup does not exist, and we'd like to avoid that.

This should fix an occasional failure like this one:

	not ok 84 kill detached busybox
	# (in test file tests/integration/kill.bats, line 27)
	#   `[ "$status" -eq 0 ]' failed
	....
	# runc kill test_busybox KILL (status=0):
	# runc kill -a test_busybox 0 (status=1):
	# time="2023-04-04T18:24:27Z" level=error msg="lstat /sys/fs/cgroup/devices/system.slice/runc-test_busybox.scope: no such file or directory"

Fixes: #3372
Fixes: #3744

### CODE REVIEW

**Review by AkihiroSuda — APPROVED** [review]

**Comment by kolyshkin** [comment]
@lifubang @thaJeztah PTAL

File: libcontainer/container_linux.go [hunk 1]
```diff
@@ -112,28 +112,33 @@ func (c *Container) OCIState() (*specs.State, error) {
...
 func (c *Container) Processes() ([]int, error) {
-	var pids []int
-	status, err := c.currentStatus()
-	if err != nil {
-		return pids, err
-	}
-	// for systemd cgroup, the unit's cgroup path will be auto removed if container's all processes exited
-	if status == Stopped && !c.cgroupManager.Exists() {
+	pids, err := c.cgroupManager.GetAllPids()
+	if c.ignoreCgroupError(err) == nil {
```
Comment (lifubang) [inline libcontainer/container_linux.go:138]: Maybe there is a more readable implement like this:
1. change the return type of `ignoreCgroupError` to `bool`;
2. if err == nil || c.ignoreCgroupError(err) {
Otherwise the reader needs to see the function body of `ignoreCgroupError` to find out what happened here.

**Comment by lifubang** [comment]
Yes, I think there is a race condition here, because `systemd` is a service, we can't know the exactly time of the removal of the cgroup path is gone when there is no pids in this cgroup.

File: libcontainer/container_linux.go [hunk 2]
```diff
@@ -112,28 +112,33 @@ func (c *Container) OCIState() (*specs.State, error) {
...
+		return nil
+	}
+	if !errors.Is(err, os.ErrNotExist) {
+		return err
+	}
+	if c.runType() == Stopped && !c.cgroupManager.Exists() {
+		return nil
+	}
+	return err
+}
```
Comment (thaJeztah) [inline libcontainer/container_linux.go:128]: Jus double checking, as the logic is a bit complex.

If the error _is_ an `os.errNotExist`, then we need to check the second part? (`if c.runType() == Stopped && !c.cgroupManager.Exists() {`) is that correct? (making sure the intent here was not `ErrNotExist` errors can be ignored).

Could flip the logic here as well to allow early return, and skipping `c.cgroupManager.Exists()`;

```go
if c.runType() != Stopped || c.cgroupManager.Exists() {
	return err
}

return nil
```

If the intent was to ignore `os.ErrNotExist`, then the first part could be;


```go
func (c *Container) ignoreCgroupError(err error) error {
 	if err == nil || errors.Is(err, os.ErrNotExist) {
 		return nil
 	}
```

File: libcontainer/container_linux.go [hunk 3]
```diff
@@ -112,28 +112,33 @@ func (c *Container) OCIState() (*specs.State, error) {
...
 // unless the container state is PAUSED in which case every PID in the slice is
 // valid.
 func (c *Container) Processes() ([]int, error) {
-	var pids []int
-	status, err := c.currentStatus()
-	if err != nil {
-		return pids, err
-	}
-	// for systemd cgroup, the unit's cgroup path will be auto removed if container's all processes exited
-	if status == Stopped && !c.cgroupManager.Exists() {
```
Comment (thaJeztah) [inline libcontainer/container_linux.go:128]: In case relevant, and that code also needs a similar fix; note that there's some _very similar_ logic in `container.Signal()`; https://github.com/opencontainers/runc/blob/941e5924efdd5b42c06e1915824f674a3e831bfe/libcontainer/container_linux.go#L358-L369

File: libcontainer/container_linux.go (hunk 1 above)
Comment (kolyshkin) [reply libcontainer/container_linux.go:138]: Oh well, my initial implementation was called `isIgnorableCgroupError` and returned `bool`. Then I took the approach used by `func ignoreTerminateErrors`.

Either way is fine with me.

File: libcontainer/container_linux.go (hunk 3 above)
Comment (kolyshkin) [reply libcontainer/container_linux.go:128]: Check the commit description, I explain why I'm doing it the way I'm doing in there.

Comment (thaJeztah) [reply libcontainer/container_linux.go:128]: ❤️ sorry, should've done so (was looking in between other things 🥹🙈)

File: libcontainer/container_linux.go (hunk 2 above)
Comment (kolyshkin) [reply libcontainer/container_linux.go]: I modified the code to be (hopefully) more readable; PTAL

File: libcontainer/container_linux.go (hunk 3 above)
Comment (kolyshkin) [reply libcontainer/container_linux.go:128]: Added a note to the code why we do the same/similar check both before and after.

**Comment by kolyshkin** [comment]
Now, here'a an interesting question: why `runc kill <CTID> <SIG>` fails if the container is stopped, while `runc kill -a <CTID> <SIG>` does not?

**Comment by lifubang** [comment]
> Now, here'a an interesting question: why `runc kill <CTID> <SIG>` fails if the container is stopped, while `runc kill -a <CTID> <SIG>` does not?

I think this happened when randomly hitting this issue.
Both `kill` with `-a` or not may hit this issue randomly. Because `ps` and `kill` with `-a` are all calling `m.GetAllPids`.

**Comment by kolyshkin** [comment]
> > Now, here'a an interesting question: why `runc kill <CTID> <SIG>` fails if the container is stopped, while `runc kill -a <CTID> <SIG>` does not?
> 
> I think this happened when randomly hitting this issue. Both `kill` with `-a` or not may hit this issue randomly. Because `ps` and `kill` with `-a` are all calling `m.GetAllPids`.

Let me rephrase my question.

Currently, on a stopped container, `runc kill` fails, but `runc kill -a` does not. See:

```console
[root@kir-rhat runc-tst]# ./runc list
ID          PID         STATUS      BUNDLE                                                CREATED                          OWNER
123         0           stopped     /home/kir/go/src/github.com/opencontainers/runc-tst   2023-04-06T21:51:20.521334579Z   root
[root@kir-rhat runc-tst]# ./runc kill 123; echo $?
ERRO[0000] container not running                        
1
[root@kir-rhat runc-tst]# ./runc kill -a 123; echo $?
0
```

My question was -- is this (the fact that adding `-a` makes the error go away) an intended behavior, and why?

**Comment by lifubang** [comment]
> My question was -- is this (the fact that adding `-a` makes the error go away) an intended behavior, and why?

As described in `runtime-spec`, please see https://github.com/opencontainers/runtime-spec/blob/main/runtime.md?plain=1#L131-L137 
`Attempting to send a signal to a container that is neither created nor running MUST have no effect on the container and MUST generate an error.`
I think this is an intended behavior.

For `-a`, it's not defined in `runtime-spec`. But I think it's also right for the stopped container, for example: shared namespace containers. So, shall we need to add a description for `-a` in `runtime-spec`?

**Comment by kolyshkin** [comment]
> For `-a`, it's not defined in `runtime-spec`. But I think it's also right for the stopped container, for example: shared namespace containers. So, shall we need to add a description for `-a` in `runtime-spec`?

Good question. So, it looks that `kill -a` currently violates the spec.

There are a few ways to look at how `runc kill -a` should work:

1. Adhere to runtime-spec description quoted above, and return a "container not running" error (i.e. same behavior as without `-a` option).
2. Amend the `-a` description with "Ignore the container-not-running error". This makes it possible to use the command in a scenario where we want to kill and destroy a container no matter what.
3. Add `--ignore-stopped` boolean option to achieve the same scenario as in 2 (and optionally have `-a` implying `--ignore-stopped`, which can be reverted by specifying `--ignore-stopped=false` or smth).

**Comment by kolyshkin** [comment]
@giuseppe WDYT? (see previous comment)

**Comment by giuseppe** [comment]
I've no strong opinion, I've followed the same runc behavior in crun.

I am afraid to change the current behavior if anyone depends on it (although the entire mechanism seems like a big race condition).

`-a` is useful only when there is no PID namespace, since otherwise it is fine to just kill the first process.  If `-a` would follow the runtime spec and fail when the container is stopped, then it wouldn't be able to terminate the other processes.  Perhaps the behavior is fine and should just be documented?

**Comment by kolyshkin** [comment]
OK let's agree to merely document the existing behavior of `-a`.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
> OK let's agree to merely document the existing behavior of `-a`.

Are you planning to open a PR for that, or do we need a tracking ticket?

(looks like this PR may need a rebase, as it's marked "outdated")

**Comment by kolyshkin** [comment]
> Are you planning to open a PR for that, or do we need a tracking ticket?

Opened https://github.com/opencontainers/runc/pull/3834

> (looks like this PR may need a rebase, as it's marked "outdated")

Rebased

**Review by AkihiroSuda — APPROVED** [review]

**Review by leonardoadame — COMMENTED** [review]
Files change
